code/modules/integrated_electronics/core/pins.dm

```diff
@@ -0,0 +1,197 @@
+where as PULSE_CHANNEL causes circuits to work() when their pulse hits them.
+
+
+A visualization of how pins work is below.  Imagine the below image involves an addition circuit.
+When the bottom pin, the activator, receives a pulse, all the numbers on the left (input) get added, and the answer goes on the right side (output).
+
+Inputs      Outputs
+
+A [2]\      /[8] result
+B [1]-\|++|/
+C [4]-/|++|
+D [1]/  ||
+        ||
+     Activator
+
+
+
+*/
+/datum/integrated_io
+	var/name = "input/output"
+	var/obj/item/integrated_circuit/holder
+	var/datum/weakref/data  // This is a weakref, to reduce typecasts.  Note that oftentimes numbers and text may also occupy this.
+	var/list/linked = list()
+	var/io_type = DATA_CHANNEL
+
+/datum/integrated_io/New(newloc, name1, new_data)
+	name = name1
+	if(new_data)
+		data = new_data
+	holder = newloc
+	if(!istype(holder))
+		message_admins("ERROR: An integrated_io ([name]) spawned without a valid holder!  This is a bug.")
+
+/datum/integrated_io/Destroy()
+	disconnect()
+	data = null
+	holder = null
+	return ..()
+/*
+/datum/integrated_io/nano_host()
+	return holder.nano_host()
+*/
+
+/datum/integrated_io/proc/data_as_type(var/as_type)
+	if(!isweakref(data))
+		return
+	var/datum/weakref/w = data
+	var/output = w.resolve()
+	return istype(output, as_type) ? output : null
+
+/datum/integrated_io/proc/display_data(var/input)
+	if(isnull(input))
+		return "(null)" // Empty data means nothing to show.
+
+	if(istext(input))
+		return "(\"[input]\")" // Wraps the 'string' in escaped quotes, so that people know it's a 'string'.
+
+/*
+list[](
+	"A",
+	"B",
+	"C"
+)
+*/
+
+	if(islist(input))
+		var/list/my_list = input
+		var/result = "list\[[my_list.len]\]("
+		if(my_list.len)
+			result += "<br>"
+			var/pos = 0
+			for(var/line in my_list)
+				result += "[display_data(line)]"
+				pos++
+				if(pos != my_list.len)
+					result += ",<br>"
+			result += "<br>"
+		result += ")"
+		return result
+
+	if(isweakref(input))
+		var/datum/weakref/w = input
+		var/atom/A = w.resolve()
+		return A ? "([A.name] \[Ref\])" : "(null)" // For refs, we want just the name displayed.
+		//return A ? "([REF(A)] \[Ref\])" : "(null)"
+
+	return "([input])" // Nothing special needed for numbers or other stuff.
+
+/datum/integrated_io/activate/display_data()
+	return "(\[pulse\])"
+
+/datum/integrated_io/proc/display_pin_type()
+	return IC_FORMAT_ANY
+
+/datum/integrated_io/activate/display_pin_type()
+	return IC_FORMAT_PULSE
+
+/datum/integrated_io/proc/scramble()
+	if(isnull(data))
+		return
+	if(isnum(data))
+		write_data_to_pin(rand(-10000, 10000))
+	if(istext(data))
+		write_data_to_pin("ERROR")
+	push_data()
+
+/datum/integrated_io/activate/scramble()
+	push_data()
+
+/datum/integrated_io/proc/write_data_to_pin(var/new_data)
+	if(isnull(new_data) || isnum(new_data) || istext(new_data) || isweakref(new_data)) // Anything else is a type we don't want.
```

Comment:
Adding an islist(new_data) here lets <ANY> pins take lists, which allows stuff like memory to store lists

Reply:
already have.But git is glitching